libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -370,28 +404,36 @@ instance
          in trace
               (raiseMockError lastSlot nextSlotNo epochstate pdfs txsl gs)
               (error "FAILS")
-      -- Left pdfs -> trace ("Discard\n"++show(ppList (ppMockChainFailure reify) pdfs)) discard -- TODO should we enable this?
       Right mcs2 -> seq mcs2 (pure mockblock)
 
-  shrinkSignal (MockBlock _ _ xs) | SS.null xs = []
-  shrinkSignal (MockBlock i s xs) = [MockBlock i s (SS.drop 1 xs), MockBlock i s (SS.take (SS.length xs - 1) xs)]
+  shrinkSignal _ = []
 
-mapProportion :: (v -> Int) -> Map.Map k v -> Gen k
-mapProportion toInt m =
+--  shrinkSignal (MockBlock _ _ xs) | SS.null xs = []
+--  shrinkSignal (MockBlock i s xs) = [MockBlock i s (SS.drop 1 xs), MockBlock i s (SS.take (SS.length xs - 1) xs)]
+
+mapProportion :: (EpochNo, Word64, Int) -> (v -> Int) -> Map.Map k v -> Gen k
+mapProportion (epochnum, lastSlot, count) toInt m =
   if null pairs
-    then error "There are no stakepools to choose an issuer from"
+    then
+      error
+        ( "There are no stakepools to choose an issuer from"
+            ++ "\n  epoch = "
+            ++ show epochnum
+            ++ "\n  last slot ="
+            ++ show lastSlot
+            ++ "\n  index of Tx in the trace = "
+            ++ show count
+        )
     else
       if all (\(n, _k) -> n == 0) pairs
-        then
-          trace
-            "All stakepools have zero stake, choose issuer arbitrarily. Probably caused by epoch boundary issue."
-            (snd (head pairs))
-        else frequency pairs
+        then snd (head pairs)
+        else -- All stakepools have zero Stake, choose issuer arbitrarily. possible, but rare.
+          frequency pairs
   where
     pairs = [(toInt v, pure k) | (k, v) <- Map.toList m]
 
-chooseIssuer :: PoolDistr crypto -> Gen (KeyHash 'StakePool crypto)
-chooseIssuer (PoolDistr m) = mapProportion getInt m
+chooseIssuer :: (EpochNo, Word64, Int) -> PoolDistr crypto -> Gen (KeyHash 'StakePool crypto)
+chooseIssuer triple (PoolDistr m) = mapProportion triple getInt m
```

Comment:
It is called overfunctionalization when a function is being passed to a function to just be applied within. It should be:
```suggestion
chooseIssuer triple (PoolDistr m) = mapProportion triple (getInt <$> m)
```

Reply:
done